pwiz_tools/Skyline/Model/Proteome/ProteinAssociation.cs

```diff
@@ -561,11 +610,65 @@ public void ApplyParsimonyOptions(bool groupProteins, bool findMinimalProteinLis
             _finalResults.FinalSharedPeptideCount = sharedPeptidesRemaining.Values.Sum();
         }
 
-        private Dictionary<IProteinRecord, PeptideAssociationGroup> CalculateProteinGroups(MappingResultsInternal results, ILongWaitBroker broker)
+        public class GeneLevelEqualityComparer : EqualityComparer<IProteinRecord>
+        {
+            public override bool Equals(IProteinRecord x, IProteinRecord y)
+            {
+                if (ReferenceEquals(x, y)) return true;
+                if (ReferenceEquals(null, x)) return false;
+                if (ReferenceEquals(null, y)) return false;
+                if (x.Metadata.Gene.IsNullOrEmpty() && y.Metadata.Gene.IsNullOrEmpty()) return x.Sequence.Equals(y.Sequence);
+                if (x.Metadata.Gene.IsNullOrEmpty() != y.Metadata.Gene.IsNullOrEmpty()) return false;
+
+                return x.Metadata.Gene.Equals(y.Metadata.Gene);
+            }
+
+            public override int GetHashCode(IProteinRecord obj)
+            {
+                if (obj.Metadata.Gene == null)
+                    return obj.Sequence.GetHashCode();
+                return obj.Metadata.Gene.GetHashCode();
+            }
+        }
+
+        private Dictionary<IProteinRecord, PeptideAssociationGroup> CalculateProteinOrGeneGroups(MappingResultsInternal results, bool geneLevel, ILongWaitBroker broker)
         {
             var _peptideGroupToProteins = new Dictionary<PeptideAssociationGroup, List<IProteinRecord>>();
 
-            foreach(var kvp in AssociatedProteins)
+            var proteinOrGeneToPeptideGroup = AssociatedProteins;
+            if (geneLevel)
+            {
+                // gene to protein to peptides; the top level dictionary uses the GeneLevelEqualityComparer
+                var geneToPeptides = new Dictionary<IProteinRecord, Dictionary<IProteinRecord, List<PeptideDocNode>>>(new GeneLevelEqualityComparer());
+                foreach (var kvp in AssociatedProteins)
+                    if (!geneToPeptides.ContainsKey(kvp.Key))
+                        geneToPeptides.Add(kvp.Key, new Dictionary<IProteinRecord, List<PeptideDocNode>> { { kvp.Key, kvp.Value.Peptides } });
+                    else
+                        geneToPeptides[kvp.Key][kvp.Key] = kvp.Value.Peptides;
+                // now pick the protein with the longest sequence if it contains all the peptides, or a concatenation of all of the sequences if not
+                IProteinRecord GenerateConcatenatedSequenceIfNecessary(Dictionary<IProteinRecord, List<PeptideDocNode>> proteinToPeptides)
+                {
+                    if (proteinToPeptides.Count == 1)
+                        return proteinToPeptides.Keys.First();
+
+                    var longestProtein = proteinToPeptides.OrderByDescending(kvp2 => kvp2.Key.Sequence.Sequence.Length).First().Key;
+                    var allPeptides = proteinToPeptides.Values.SelectMany(o => o).Distinct().ToList();

```

Comment:
Oops, I missed this when I was reviewing your reversion of my overzealous ReferenceValue refactoring. But it seems I wouldn't need to change these usages if we kept my overzealous approach, right? It might just be simpler to use ReferenceValue<PeptideDocNode> *everywhere* in ProteinAssociation with an explicit comment that it's to "simplify" the code and make it clear the nodes are almost never mutated.